### PR TITLE
test utilities: easy way to simulate terminal context

### PR DESCRIPTION
**Pull request by cre4ture**
addresses #5785
and addresses #1857
replaces if accepted PR #5858

let me know what you think. It's very easy to use. Just do a `.simulate_terminal(true)` and thats it.
I implemented in such a way that also stdin piping still works and also the capturing of stdout and stderr works as usual.
EDIT: For cases where the terminal size matters, there is no an additional function available:
```
.terminal_size(libc::winsize {
                ws_col: 40,
                ws_row: 10,
                ws_xpixel: 40 * 8,
                ws_ypixel: 10 * 10,
            })
```

Example:

```rust
#[test]
fn test_simulation_of_terminal_true() {
    let scene = TestScenario::new(util_name!());

    let out = scene
        .ucmd()
        .arg("sh")
        .arg("is_atty.sh")
        .terminal_simulation(true)
        .succeeds();
    assert_eq!(
        String::from_utf8_lossy(out.stdout()),
        "stdin is atty\r\nstdout is atty\r\nstderr is atty\r\n"
    );
    assert_eq!(
        String::from_utf8_lossy(out.stderr()),
        "This is an error message.\r\n"
    );
}
```

where `is_atty.sh` looks like this:

```bash
#!/bin/bash

if [ -t 0 ] ; then
    echo "stdin is atty"
else
    echo "stdin is not atty"
fi

if [ -t 1 ] ; then
    echo "stdout is atty"
else
    echo "stdout is not atty"
fi

if [ -t 2 ] ; then
    echo "stderr is atty"
else
    echo "stderr is not atty"
fi

echo "This is an error message." > /dev/stderr

true

```

### CODE REVIEW

File: tests/common/util.rs
Comment (RenjiSann) [inline tests/common/util.rs:2371]: Is there a specific purpose here to creating 2 temporary result values ?

Comment (cre4ture) [reply tests/common/util.rs:2371]: Thanks for pointing out. I was debugging this part. Thats why I added the intermediate result values. But this can be removed. I'll do it.

**Comment by RenjiSann** [comment]
Is it ready to merge already, or are we waiting for further improvements ?

**Comment by cre4ture** [comment]
What is missing is a review and approval from one of the maintainers.
@sylvestre - I did also the suash commit now.

File: .gitignore
Comment (tertsdiepraam) [inline .gitignore:17]: What is this for?

File: tests/by-util/test_env.rs
Comment (tertsdiepraam) [inline tests/by-util/test_env.rs:254]: Are these meaningful tests of `env`? If not, it could still be useful to have them (we're going very meta with tests for our test framework 😄), but maybe somewhere else?

**Review by tertsdiepraam — APPROVED** [review]
I like it a lot! Excellent work! I only have some small questions.

I also think that we might want to be able to configure the terminal with this interface, but for now, the API is flexible enough I think. It's also about time we extract the test framework to a crate, so we can generate documentation for it. It's getting quite complex. That's something for another PR though.

I would like to wait for approval of one of the other maintainers before merging though.

One final question: have you seen any intermittently failing tests with this? Can we assume that this implementation is robust enough not to make our test suite fail unexpectedly (even more so then it already does 😅)

File: .gitignore
Comment (cre4ture) [reply .gitignore:17]: I started using `git worktree`s. E.g. a seperate worktree for a clean copy of the main branch such that I can do cherry picking from my messy develop branch and push small PRs.

I prefix the worktree with `worktree` and to avoid that this pops up in the diff, I put it in .gitignore. I hope you don't mind.

File: tests/by-util/test_env.rs
Comment (cre4ture) [reply tests/by-util/test_env.rs:254]: ah, yes. This is a valid point. I used `env` here as a tool rather than to test itself. This is something I should cleanup. Do you have a proposal for a better location of these tests?

File: .gitignore
Comment (tertsdiepraam) [reply .gitignore:17]: That should go in your own personal git configuration. The project `.gitignore` is not meant to enable your personal workflow 😄

File: tests/by-util/test_env.rs
Comment (tertsdiepraam) [reply tests/by-util/test_env.rs:254]: Maybe in the file where the test framework is defined? It's an interesting question, we don't really have tests for the tests yet 😄

Comment (cre4ture) [reply tests/by-util/test_env.rs]: done

File: .gitignore
Comment (cre4ture) [reply .gitignore]: thanks for the hint. I never considered this.

**Comment by cre4ture** [comment]
> One final question: have you seen any intermittently failing tests with this? Can we assume that this implementation is robust enough not to make our test suite fail unexpectedly (even more so then it already does 😅)

I think its stable.

**Comment by sylvestre** [comment]
thanks

**Pull request merged**
